File: js/agents/src/patterns/handoff.ts
import type { AgentRoom } from "../AgentRoom";
import type { TaskEnvelope, ResultEnvelope } from "../types";
import { CorrelationManager } from "../correlation";
import { createTaskEnvelope, createResultEnvelope } from "../envelope";
import { IncompatibleProtocolError } from "../errors";

type TaskHandler = (
  task: TaskEnvelope,
  respond: (
    status: ResultEnvelope["status"],
    payload: Record<string, unknown>,
    error?: { code: string; message: string },
  ) => void,
) => void;

/**
 * Handoff pattern — dispatch tasks to agents and receive results.
 *
 * Orchestrators use `dispatch()` to send work. The SDK checks if any
 * connected agent has the requested capability (via presence-based
 * service discovery) before dispatching.
 *
 * Workers use `onTask()` with a capabilities filter — they only receive
 * tasks matching their registered capabilities.
 *
 * @example
 * ```typescript
 * // Orchestrator
 * const handoff = new Handoff(room);
 * const result = await handoff.dispatch('summarize', { text }, { waitForResult: true });
 *
 * // Worker
 * const handoff = new Handoff(room);
 * handoff.onTask(['summarize', 'translate'], async (task, respond) => {
 *   const output = await processTask(task);
 *   respond('success', { output });
 * });
 * ```
 */
export class Handoff {
  private _room: AgentRoom;
  private _correlations = new CorrelationManager<ResultEnvelope>();
  private _warnedMixed = false;

  constructor(room: AgentRoom) {
    this._room = room;

    // Wire result correlation
    this._room.on("result", (envelope) => {
      this._correlations.resolve(envelope.correlationId, envelope);
    });
  }

  /**
   * Dispatch a task to agents with the given capability.
   *
   * Uses presence-based service discovery to verify at least one agent
   * can handle the capability before dispatching. Throws if no capable
   * agent is connected (unless `allowNoWorkers` is set).
   */
  async dispatch(
    capability: string,
    payload: Record<string, unknown>,
    options?: {
      tags?: string[];
      priority?: TaskEnvelope["priority"];
      timeout?: number;
      waitForResult?: boolean;
      metadata?: Record<string, unknown>;
      /** Attribution override; defaults to the room's agentId */
      createdBy?: string;
      /** Reply address override; defaults to the room's agentId (its results filter) */
      replyTo?: string;
      /** Skip the capability check (dispatch even if no workers are connected) */
      allowNoWorkers?: boolean;
      /**
       * Persistent Presence: require an ONLINE capable agent. By default the gate
       * also accepts an offline persistent agent (discoverable, and woken by the
       * broker on publish) — set this to restore strict, low-latency behaviour.
       */
      requireOnline?: boolean;
      /** Skip the protocol fail-fast (responders on 0.2.x have directed replies but don't advertise protocol yet) */
      allowLegacyResponders?: boolean;
    },
  ): Promise<ResultEnvelope | void> {
    // Service discovery: check if any agent can handle this capability.
    // Persistent Presence: findAgents includes offline persistent agents, which
    // the broker wakes on publish — so they satisfy the gate unless requireOnline.
    if (!options?.allowNoWorkers) {
      const capable = this._room.findAgents(capability);
      const usable = options?.requireOnline
        ? capable.filter((a) => a.status === undefined || a.status === "online")
        : capable;
      if (usable.length === 0) {
        throw new Error(
          `No ${options?.requireOnline ? "online " : ""}agent with capability "${capability}" is available. ` +
          `Available capabilities: [${this._room.getAvailableCapabilities().join(', ')}]. ` +
          `Connected agents: ${this._room.getConnectedAgents().length}. ` +
          `Use { allowNoWorkers: true } to dispatch anyway.`
        );
      }
    }

    const envelope = createTaskEnvelope(capability, payload, {
      ...options,
      createdBy: options?.createdBy ?? this._room.agentId,
      // Reply address: workers publish the result filter-directed to this
      // room's results subscription
      replyTo: options?.replyTo ?? this._room.agentId,
    });
    this._room.publishTask(envelope);

    if (options?.waitForResult) {
      // Fail fast when the outcome is deterministic: if capable workers are
      // visible and ALL advertise agents-protocol < 2, their results cannot
      // reach this dispatcher's filtered subscription. Mixed pools proceed
      // with a warning (presence is eventually consistent).
      const capable = this._room.findAgents(capability);
      if (!options?.allowLegacyResponders && capable.length > 0) {
        const modern = capable.filter((a) => a.protocol >= 2);
        if (modern.length === 0) {
          throw new IncompatibleProtocolError(
            `Task '${capability}' dispatch with waitForResult`,
            capable.map((a) => ({ name: a.name, protocol: a.protocol })),
          );
        }
        if (modern.length < capable.length && !this._warnedMixed) {
          this._warnedMixed = true;
          console.warn(
            `[nolag-agents] Capability '${capability}' has workers on agents-protocol < 2: ` +
              capable.filter((a) => a.protocol < 2).map((a) => a.name).join(", ") +
              ". Their results may not be delivered — upgrade them.",
          );
        }
      }

      return this._correlations.register(
        envelope.correlationId,
        options.timeout,
        `Task '${capability}' dispatch (${capable.length} capable worker${capable.length === 1 ? "" : "s"} visible). ` +
          `Likely causes: worker crashed mid-task, worker on agents-protocol < 2 ` +
          `(results not directed), or the room is not deliverable`,
      );
    }
  }

  /**
   * Register a handler for incoming tasks, filtered by capabilities.
   *
   * Only tasks whose `capability` field matches one of the provided
   * capabilities will be delivered to the handler. Non-matching tasks
   * are silently ignored.
   *
   * @param capabilities - Array of capabilities this worker handles.
   *                       Pass `'*'` to receive all tasks.
   * @param handler - Async handler called with the task and a respond function.
   */
  onTask(handler: TaskHandler): void;
  onTask(capabilities: string[] | '*', handler: TaskHandler): void;
  onTask(
    capabilitiesOrHandler: string[] | '*' | TaskHandler,
    maybeHandler?: TaskHandler,
  ): void {
    // Single-arg form: onTask(handler) receives all tasks
    const capabilities: string[] | '*' =
      typeof capabilitiesOrHandler === "function" ? '*' : capabilitiesOrHandler;
    const handler: TaskHandler =
      typeof capabilitiesOrHandler === "function" ? capabilitiesOrHandler : maybeHandler!;

    this._room.on("task", (task) => {
      // Filter by capability unless wildcard
      if (capabilities !== '*' && !capabilities.includes(task.capability)) {
        return;
      }

      const respond = (
        status: ResultEnvelope["status"],
        payload: Record<string, unknown>,
        error?: { code: string; message: string },
      ) => {
        const result = createResultEnvelope(
          task.taskId,
          task.correlationId,
          status,
          payload,
          error,
          this._room.agentId,
          // Direct the result to the dispatcher's filter sub-topic
          task.replyTo ?? task.createdBy,
        );
        this._room.publishResult(result);
      };
      handler(task, respond);
    });
  }

  /**
   * Get agents capable of handling a specific task type.
   * Delegates to the room's presence-based service discovery.
   */
  getCapableAgents(capability: string) {
    return this._room.findAgents(capability);
  }

  /** Cancel all pending correlations */
  dispose(): void {
    this._correlations.clear();
  }
}

File: python/agents/nolag_agents/patterns/handoff.py
from __future__ import annotations

from typing import Any, Callable, Optional, Union

from ..agent_room import AgentRoom
from ..types import TaskEnvelope, ResultEnvelope, from_camel_dict
from ..correlation import CorrelationManager
from ..envelope import create_task_envelope, create_result_envelope
from ..errors import IncompatibleProtocolError


class Handoff:
    """Task dispatch and capability routing pattern.

    Orchestrators use dispatch() to send work. Workers use on_task()
    with a capabilities filter.
    """

    def __init__(self, room: AgentRoom) -> None:
        self._room = room
        self._correlations: CorrelationManager[ResultEnvelope] = CorrelationManager()
        self._warned_mixed = False

        self._room.on("result", self._on_result)

    def _on_result(self, data: Any) -> None:
        d = data if isinstance(data, dict) else {}
        correlation_id = d.get("correlationId", d.get("correlation_id", ""))
        if correlation_id:
            envelope = _dict_to_result(d)
            self._correlations.resolve(correlation_id, envelope)

    async def dispatch(
        self,
        capability: str,
        payload: dict[str, Any],
        *,
        tags: Optional[list[str]] = None,
        priority: str = "medium",
        timeout: Optional[int] = None,
        wait_for_result: bool = False,
        metadata: Optional[dict[str, Any]] = None,
        allow_no_workers: bool = False,
        allow_legacy_responders: bool = False,
    ) -> Optional[ResultEnvelope]:
        if not allow_no_workers:
            capable = self._room.find_agents(capability)
            if len(capable) == 0:
                available = self._room.get_available_capabilities()
                connected = self._room.get_connected_agents()
                raise RuntimeError(
                    f'No agent with capability "{capability}" is connected. '
                    f"Available capabilities: [{', '.join(available)}]. "
                    f"Connected agents: {len(connected)}. "
                    f"Use allow_no_workers=True to dispatch anyway."
                )

        envelope = create_task_envelope(
            capability,
            payload,
            tags=tags,
            priority=priority,
            timeout=timeout,
            metadata=metadata,
            created_by=self._room.agent_id,
            # Reply address: workers publish the result filter-directed to
            # this room's results subscription
            reply_to=self._room.agent_id,
        )
        await self._room.publish_task(envelope)

        if wait_for_result:
            # Fail fast when the outcome is deterministic: if capable workers
            # are visible and ALL advertise agents-protocol < 2, their results
            # cannot reach this dispatcher's filtered subscription. Mixed
            # pools proceed with a warning (presence is eventually consistent).
            capable = self._room.find_agents(capability)
            if not allow_legacy_responders and capable:
                modern = [a for a in capable if a.protocol >= 2]
                if not modern:
                    raise IncompatibleProtocolError(
                        f"Task '{capability}' dispatch with wait_for_result",
                        [(a.name, a.protocol) for a in capable],
                    )
                if len(modern) < len(capable) and not self._warned_mixed:
                    self._warned_mixed = True
                    import logging
                    logging.getLogger("nolag_agents").warning(
                        "Capability '%s' has workers on agents-protocol < 2: %s — "
                        "their results may not be delivered; upgrade them.",
                        capability,
                        ", ".join(a.name for a in capable if a.protocol < 2),
                    )

            n = len(capable)
            return await self._correlations.register(
                envelope.correlation_id,
                timeout,
                f"Task '{capability}' dispatch ({n} capable worker{'' if n == 1 else 's'} "
                f"visible). Likely causes: worker crashed mid-task, worker on "
                f"agents-protocol < 2 (results not directed), or the room is "
                f"not deliverable",
            )
        return None

    def on_task(
        self,
        capabilities: Union[list[str], str],
        handler: Callable[[TaskEnvelope, Callable[..., Any]], Any],
    ) -> None:
        def _handler(data: Any) -> None:
            d = data if isinstance(data, dict) else {}
            task = _dict_to_task(d)

            if capabilities != "*" and task.capability not in capabilities:
                return

            async def _respond(
                status: str,
                result_payload: dict[str, Any],
                error: Optional[dict[str, str]] = None,
            ) -> None:
                result = create_result_envelope(
                    task.task_id,
                    task.correlation_id,
                    status,
                    result_payload,
                    error,
                    self._room.agent_id,
                    # Direct the result to the dispatcher's filter sub-topic
                    reply_to=task.reply_to or task.created_by,
                )
                await self._room.publish_result(result)

            handler(task, _respond)

        self._room.on("task", _handler)

    def get_capable_agents(self, capability: str):
        return self._room.find_agents(capability)

    def dispose(self) -> None:
        self._correlations.clear()


def _dict_to_task(d: dict[str, Any]) -> TaskEnvelope:
    return TaskEnvelope(
        type="task",
        task_id=d.get("taskId", d.get("task_id", "")),
        correlation_id=d.get("correlationId", d.get("correlation_id", "")),
        capability=d.get("capability", ""),
        priority=d.get("priority", "medium"),
        payload=d.get("payload", {}),
        reply_to=d.get("replyTo", d.get("reply_to")),
        tags=d.get("tags"),
        metadata=d.get("metadata"),
        created_at=d.get("createdAt", d.get("created_at", 0)),
        created_by=d.get("createdBy", d.get("created_by")),
        timeout=d.get("timeout"),
    )


def _dict_to_result(d: dict[str, Any]) -> ResultEnvelope:
    return ResultEnvelope(
        type="result",
        task_id=d.get("taskId", d.get("task_id", "")),
        correlation_id=d.get("correlationId", d.get("correlation_id", "")),
        status=d.get("status", "success"),
        payload=d.get("payload", {}),
        error=d.get("error"),
        completed_at=d.get("completedAt", d.get("completed_at", 0)),
        completed_by=d.get("completedBy", d.get("completed_by")),
        reply_to=d.get("replyTo", d.get("reply_to")),
    )
